frontend/catalyst/utils/jax_extras.py

```diff
@@ -221,75 +264,132 @@ def __init__(self, boxid: int, e: JaxprEqn):
     return [b.e for b in stable_toposort(boxes)]  # [4]
 
 
-def initial_style_jaxprs_with_common_consts1(
-    funs: Sequence[Callable], in_tree, in_avals, primitive_name: str
-):
-    """This function is the head (shorter) part of the original
-    `lax.control_flow.common._initial_style_jaxprs_with_common_consts` of JAX. The algorithm is the
-    same at the time of this writing, we use this function only to avoid conflicts with future
-    versions of JAX.
-    """
-    jaxprs, all_consts, all_out_trees = unzip3(
-        _initial_style_open_jaxpr(fun, in_tree, in_avals, primitive_name) for fun in funs
+def jaxpr_pad_consts(jaxprs: List[Jaxpr]) -> List[ClosedJaxpr]:
+    """Align the constants of Jaxpr programs. Return the list of corresponding programs accepting
+    same constants."""
+    newvar = gensym(jaxprs, suffix="_")
+
+    # pylint: disable=too-many-nested-blocks
+    all_padded_constvars = []
+    for jaxpr in jaxprs:
```

Comment:
There is a 4x times nested loop, is there anyway to simplify that?

Reply:
Indeed, we might reduce the number of loops by 1, thanks! Here we make every conditional branch (a Jaxpr program) to contain all constants found in all conditional branches. Without optimizations, we would get $O(B^2 * C * D)$ complexity where $B$ - number of branches, $C$ - maximum number of constant tensors, $D$ - maximum number of tensor dimensions. But we can pre-calculate the dimensions parts.